nilearn/datasets/tests/test_atlas.py

```diff
@@ -243,6 +294,130 @@ def test_fail_fetch_atlas_harvard_oxford(tmp_path, request_mocker):
+
+    atlas_data[:5, 7:9, :] = 3
+    atlas_data[5:, 7:9, :] = 3
+
+    nibabel.Nifti1Image(atlas_data, np.eye(4) * 3).to_filename(
+        target_atlas_nii)
+
+    with open(os.path.join(ho_dir, 'Juelich.xml'), 'w') as dummy:
+        dummy.write("<?xml version='1.0' encoding='us-ascii'?>\n"
+                    "<data>\n"
+                    '<label index="0" x="125" y="75" z="107">R1</label>\n'
+                    '<label index="1" x="53" y="78" z="108">R2</label>\n'
+                    '<label index="2" x="135" y="89" z="109">R3</label>\n'
+                    "</data>")
+        dummy.close()
+
+    # when symmetric_split=False (by default), then atlas fetcher should
+    # have maps as string and n_labels=4 with background.
+    # Since, we relay on xml file to retrieve labels.
+    ho = atlas.fetch_atlas_juelich(atlas_name=target_atlas,
+                                   data_dir=str(tmp_path))
+    assert isinstance(ho.maps, nibabel.Nifti1Image)
+    assert isinstance(ho.labels, np.ndarray)
+    assert len(ho.labels) == 4
+    assert ho.labels[0] == "Background"
+    assert ho.labels[1] == "R1"
+    assert ho.labels[2] == "R2"
+    assert ho.labels[3] == "R3"
+
+    split_atlas_fname = 'Juelich-maxprob-thr0-1mm.nii.gz'
+    nifti_target_split = os.path.join(nifti_dir, split_atlas_fname)
+    nibabel.Nifti1Image(atlas_data, np.eye(4) * 3).to_filename(
+        nifti_target_split)
+
+    with open(os.path.join(ho_dir, 'Juelich.xml'), 'w') as dummy:
+        dummy.write("<?xml version='1.0' encoding='us-ascii'?>\n"
+                    "<data>\n"
+                    '<label index="0" x="125" y="75" z="107">R1</label>\n'
+                    '<label index="1" x="53" y="78" z="108">R2</label>\n'
+                    '<label index="2" x="135" y="89" z="109">R3</label>\n'
+                    "</data>")
+        dummy.close()
+
+    # when symmetric_split=False (by default), then atlas fetcher should
+    # have maps as string and n_labels=4 with background.
+    # Since, we relay on xml file to retrieve labels.
+    ho_wo = atlas.fetch_atlas_juelich(atlas_name="maxprob-thr0-1mm",
+                                      data_dir=str(tmp_path))
+    assert isinstance(ho_wo.maps, nibabel.Nifti1Image)
+    assert isinstance(ho_wo.labels, np.ndarray)
+    assert len(ho_wo.labels) == 4
+    assert ho_wo.labels[0] == "Background"
+    assert ho_wo.labels[1] == "R1"
+    assert ho_wo.labels[2] == "R2"
+    assert ho_wo.labels[3] == "R3"
+
+    # This section tests with lateralized version. In other words,
+    # symmetric_split=True
+
+    # We test the fetcher with symmetric_split=True by creating a new
+    # dummy local file and fetch them and test the output variables
+    # accordingly.
+    with open(os.path.join(ho_dir, 'Juelich.xml'), 'w') as dummy2:
+        dummy2.write("<?xml version='1.0' encoding='us-ascii'?>\n"
+                     "<data>\n"
+                     '<label index="0" x="125" y="75" z="107">R1</label>\n'
```

Comment:
It looks like there is a lot of repetitions and copy-pasting in this function.
Could we factorize this? Maybe with a pytest fixture?

Reply:
Yes you are right. There was some unnecessary lines so I deleted them. Also, I refactored using ```_write_to_xml``` and ```_test_result_xml```.